Do not crash on bad annotations

seanh opened the pull request:
This is branched off of <https://github.com/hypothesis/bouncer/pull/10>. Added some defensive coding against Elasticsearch document variability. I've tested this on ~100,000 annotations from the production hypothes.is API.

chdorner commented on bouncer/views.py:
```diff
@@ -25,10 +30,27 @@ def annotation(self):
+
+        if not isinstance(document_uri, str):
+            raise httpexceptions.HTTPUnprocessableEntity(
+                _("The annotation has an invalid document URI"))
```
I've moved this comment over from #10, since it's more appropriate here.

After spending some quality time sifting through all the special ways we store annotations at the moment, I found quite a few that don't have a `uri`, and even some that don't have anything that could be used as a URI.
The Postgres migration code for example first tries to read the URI from `target[0][source]`, if that doesn't exist it tries to fall back to `uri`, and if taht one doesn't exist it refuses to migrate the annotation.

My suggestion for bouncer is something like:

1. Read from `target[0][source]`
2. When still missing or not a http(s) uri, try `uri`
3. Maybe: When still missing, find the first http(s) uri from `target[0].link` with `rel == 'canonical'` (careful, sometimes `target[0].link` is not a list but a string)
4. When all of the above fail, render an error

robertknight commented:
Is there any way we could re-use the recently written presenter logic to avoid having to deal with this variability in Bouncer itself?

nickstenning commented:
This looks ok, but pulling out a function `annotation_uri` which accepts an annotation object as input and returns a uri or `None` as output would make this _way_ clearer and a hell of a lot easier to test.

nickstenning commented:
> Is there any way we could re-use the recently written presenter logic to avoid having to deal with this variability in Bouncer itself?

As mentioned in Slack, I don't think this is worth our time. Bouncer is only reading from Elasticsearch until such time as annotation data moves to Postgres, at which point much of this won't be necessary.

seanh commented:
Gonna wait until #10 is merged before finishing this one off

seanh commented on bouncer/views.py:
```diff
@@ -25,10 +30,27 @@ def annotation(self):
+            document_uri = document["_source"]["uri"]
+        except KeyError:
+            raise httpexceptions.HTTPUnprocessableEntity(
+                _("The annotation is invalid"))
```
This should increment a statsd counter before raising

seanh replied:
Done

seanh commented on bouncer/views.py:
```diff
@@ -25,10 +30,27 @@ def annotation(self):
+                _("The annotation is invalid"))
+
+        if not isinstance(document_uri, str):
+            raise httpexceptions.HTTPUnprocessableEntity(
```
This should increment a statsd counter before raising

seanh replied:
Done

seanh commented:
Done everything except @chdorner's comment about checking `["target"][0]["source"]` and `["target"][0]["link"]` as well as `["uri"] for annotated document URIs.

Split out a few utility things from `views` into a new `util` module (I wasn't going to do this, but it turned out I had a few utils not just the one function).

Addition some missing tests for the statsd reporting.

Fixed a non-ASCII typo (from a different pull request but just fixed it here).

seanh commented:
Still need to test this manually and maybe put in the more robust annotation parsing. Lunch now..

seanh commented:
As discussed in the sprint planning meeting, let's leave out the more complex annotation parsing at least for now.

I've re-tested manually and seems to work

nickstenning commented:
LGTM.
